### PR TITLE
feat: allow basic auth for all endpoints

### PR DESCRIPTION
**Pull request by adityathebe**
Closes https://github.com/flanksource/incident-commander/issues/238
Closes https://github.com/flanksource/incident-commander/issues/257

### CODE REVIEW

File: auth/middleware.go [hunk 1]
```diff
@@ -77,6 +73,10 @@ func (k *kratosMiddleware) validateSession(r *http.Request) (*client.Session, er
 		return &login.Session, nil
 	}
 
+	if strings.HasPrefix(r.URL.Path, "/upstream_push") && !hasBasicAuth {
+		return nil, errors.New("endpoint requires basic auth")
+	}
```
Comment (adityathebe) [inline auth/middleware.go:78]: @moshloop should I remove this? So upstream can also work with cookies?

Comment (moshloop) [reply auth/middleware.go:78]: Yeh - I don't think we should different auth methods for different endpoints

File: auth/middleware.go [hunk 2]
```diff
@@ -17,17 +18,26 @@ import (
...
+	randString, err := utils.GenerateRandString(30)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate random string: %w", err)
 	}
+
+	return &kratosMiddleware{
+		client:             k.client,
+		jwtSecret:          k.jwtSecret,
+		tokenCache:         cache.New(3*24*time.Hour, 12*time.Hour),
+		authSessionCache:   cache.New(30*time.Minute, time.Hour),
```
Comment (adityathebe) [inline auth/middleware.go:38]: @moshloop I could also potentially use the same cache instance and use custom expiration time for basic auth caches.

File: auth/middleware.go [hunk 3]
```diff
@@ -57,23 +67,31 @@ func (k *kratosMiddleware) validateSession(r *http.Request) (*client.Session, er
...
 
-		username, password, ok := r.BasicAuth()
-		if !ok {
-			return nil, errors.New("endpoint requires basic auth")
-		}
-
 		updateLoginFlowBody := client.UpdateLoginFlowBody{UpdateLoginFlowWithPasswordMethod: client.NewUpdateLoginFlowWithPasswordMethod(username, "password", password)}
 		login, _, err := k.client.FrontendApi.UpdateLoginFlow(r.Context()).Flow(loginFlow.Id).UpdateLoginFlowBody(updateLoginFlowBody).Execute()
 		if err != nil {
 			return nil, fmt.Errorf("failed to update native login flow: %w", err)
```
Comment (adityathebe) [inline auth/middleware.go:88]: Even on auth failure _(when invalid credential is provided)_, I wanted to cache the response for a slightly shorter period of time _(1 minute?)_ so a faulty downstream would not be able to bombard the central instance.

 However, the error returned doesn't have a reliable way to know that the error occurred due to invalid credentials.

Comment (moshloop) [reply auth/middleware.go:88]: I think we can cache login failures for 1m irrespective of reason

Comment (adityathebe) [reply auth/middleware.go:88]: okay - I guess that's reasonable.

**Pull request merged**